pkg/query-service/app/integrations/messagingQueues/kafka/sql.go

```diff
@@ -318,6 +319,141 @@ GROUP BY
+    processed_traces AS (
+        SELECT
+            resource_string_service$$name AS service_name,
+            name AS span_name,
+            CASE
+                WHEN attribute_string_messaging$$system != '' THEN attribute_string_messaging$$system
+                WHEN (has(attributes_string, 'celery.action') OR has(attributes_string, 'celery.task_name')) THEN 'celery'
+                ELSE 'undefined'
+            END AS messaging_system,
+            kind_string,
+            COALESCE(
+                NULLIF(attributes_string['messaging.destination.name'], ''),
+                NULLIF(attributes_string['messaging.destination'], '')
+            ) AS destination,
+            durationNano,
+            status_code
+        FROM signoz_traces.distributed_signoz_index_v3
+        WHERE
+            timestamp >= toDateTime64(%f, 9)
+            AND timestamp <= toDateTime64(%f, 9)
+            AND ts_bucket_start >= toDateTime64(%f, 9)
+            AND ts_bucket_start <= toDateTime64(%f, 9)
+            AND (
+                attribute_string_messaging$$system = 'kafka'
+                OR has(attributes_string, 'celery.action')
+                OR has(attributes_string, 'celery.task_name')
+            )
+    ),
+    aggregated_metrics AS (
+        SELECT
+            service_name,
+            span_name,
+            messaging_system,
+            destination,
+            kind_string,
+            count(*) AS total_count,
+            sumIf(1, status_code = 2) AS error_count,
+            quantile(0.95)(durationNano) / 1000000 AS p95_latency -- Convert to ms
+        FROM
+            processed_traces
+        GROUP BY
+            service_name,
+            span_name,
+            messaging_system,
+            destination,
+            kind_string
+    )
+SELECT
+    aggregated_metrics.service_name,
+    aggregated_metrics.span_name,
+    aggregated_metrics.messaging_system,
+    aggregated_metrics.destination,
+    aggregated_metrics.kind_string,
+    COALESCE(aggregated_metrics.total_count / timeRange.seconds, 0) AS throughput,
+    COALESCE((aggregated_metrics.error_count * 100.0) / aggregated_metrics.total_count, 0) AS error_percentage,
+    aggregated_metrics.p95_latency
+FROM
+    aggregated_metrics
+CROSS JOIN timeRange
+WHERE
+    %s
+ORDER BY
+    aggregated_metrics.service_name,
+    aggregated_metrics.span_name;
+`,
+		timeRangeSecs, // timeRange AS (SELECT %f AS seconds)
+		startSeconds, endSeconds,
+		tsBucketStart, tsBucketEnd,
+		whereSQL,
+	)
+
+	return query
+}
+
+// inClause returns SQL like "fieldName IN ('val1','val2','val3')"
+func inClause(fieldName string, values []string) string {
+	// Quote and escape each value for safety
+	var quoted []string
+	for _, v := range values {
+		// Simple escape: replace any single quotes in v
+		safeVal := strings.ReplaceAll(v, "'", "''")
+		quoted = append(quoted, fmt.Sprintf("'%s'", safeVal))
+	}
+	return fmt.Sprintf("%s IN (%s)", fieldName, strings.Join(quoted, ","))
+}
```

Comment:
Another reason why filters from v3 should be used, then `buildTracesFilterQuery` https://github.com/SigNoz/signoz/blob/ecd50f72327db889b7665ec0701bf5ed727d7a5f/pkg/query-service/app/traces/v4/query_builder.go#L90 can help prepare the query.

Reply:
Maybe create a PR against this branch with the changes? Since this is the first PR for this feature, I suggest we get the filter changes in the same PR.